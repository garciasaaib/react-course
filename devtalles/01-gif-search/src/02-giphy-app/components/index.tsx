export { AddCategory } from './AddCategory';
export { GifGrid } from './GifGrid';
export { GifItem } from './GifItem';