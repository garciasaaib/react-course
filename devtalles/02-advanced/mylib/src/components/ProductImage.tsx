import React, { CSSProperties, useContext } from "react";
import noImage from "../assets/no-image.jpg";
import { ProductContext } from "../context/ProductContext";
import styles from "../styles/styles.module.css";

export interface ProductImageProps {
	img?: string;
	className?: string;
  style?: CSSProperties;
}
export const ProductImage = ({ img = "", className = "", style }: ProductImageProps) => {
	const { product } = useContext(ProductContext);
	return (
		<img
			src={img || product.img || noImage}
			alt="Product image"
			className={`${styles.productImg} ${className}`}
			style={style}
		/>
	);
};
