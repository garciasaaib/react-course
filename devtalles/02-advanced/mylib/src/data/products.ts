export interface Product {
	id: string;
	title: string;
	img: string;
}

export const product1 = {
	id: "1",
	title: "Coffee Mug - Card",
	img: "./coffee-mug.png",
};

export const product2 = {
	id: "2",
	title: "Coffee Mug - Meme",
	img: "./coffee-mug2.png",
};

export const products: Product[] = [product1, product2];
