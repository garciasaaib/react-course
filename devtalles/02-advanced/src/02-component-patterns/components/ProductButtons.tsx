import { CSSProperties, useContext } from "react";
import styles from "../styles/styles.module.css";
import { ProductContext } from "../context/ProductContext";

export interface ProductButtonsProps {
  className?: string;
  style?: CSSProperties;
}
export const ProductButtons = ({ className, style }: ProductButtonsProps) => {
	const { increaseBy, counter } = useContext(ProductContext);
	return (
		<div className={`${styles.buttonsContainer} ${className}`} style={style}>
			<button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
				-
			</button>
			<div className={styles.countLabel}>{counter}</div>
			<button className={styles.buttonAdd} onClick={() => increaseBy(1)}>
				+
			</button>
		</div>
	);
};