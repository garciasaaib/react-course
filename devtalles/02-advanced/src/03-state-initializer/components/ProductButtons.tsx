import { CSSProperties, useCallback, useContext } from "react";
import styles from "../styles/styles.module.css";
import { ProductContext } from "../context/ProductContext";

export interface ProductButtonsProps {
  className?: string;
  style?: CSSProperties;
}
export const ProductButtons = ({ className, style }: ProductButtonsProps) => {
	const { increaseBy, counter, maxCount } = useContext(ProductContext);

	// useCallback permite evaluar y actualizar valores sin ejecutar re-renders
	// se usa en lugar de una funcion normal cuando se manda como prop a un componente hijo
	const isMaxReached = useCallback(() => 
		!!maxCount && counter === maxCount
	, [counter, maxCount]);

	return (
		<div className={`${styles.buttonsContainer} ${className}`} style={style}>
			<button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
				-
			</button>
			<div className={styles.countLabel}>{counter}</div>
			<button className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`} onClick={() => increaseBy(1)}>
				+
			</button>
		</div>
	);
};