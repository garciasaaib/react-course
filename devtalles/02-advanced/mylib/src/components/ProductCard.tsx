import React, { CSSProperties, ReactElement } from "react";
import { useProduct } from "../hooks/useProduct";
import { Product } from "../data/products";
import { onChangeArgs } from "../hooks/useShoppingCart";
import { Provider } from "../context/ProductContext";
import { ProductCardHandlers } from "../interfaces/interfaces";
import styles from "../styles/styles.module.css";

// componente ProductCard
export interface InitialValues {
	count: number;
	maxCount: number;
}

export interface ProductCardProps {
	product: Product;
	children: (args: ProductCardHandlers) => ReactElement | ReactElement[];
	className?: string;
	style?: CSSProperties;
	onChange?: (args: onChangeArgs) => void;
	value?: number;
	initialValues?: InitialValues;
}

export const ProductCard = ({
	product,
	children,
	className,
	style,
	onChange,
	value,
	initialValues,
}: ProductCardProps) => {
	const { increaseBy, counter, maxCount, isMaxCountReached, reset } =
		useProduct({
			onChange,
			product,
			value,
			initialValues,
		});
	return (
		<Provider value={{ counter, increaseBy, product, maxCount }}>
			<div className={`${styles.productCard} ${className}`} style={style}>
				{children({
					product,
					maxCount,
					count: counter,
					isMaxCountReached,
					reset,
					increaseBy,
				})}
			</div>
		</Provider>
	);
};
