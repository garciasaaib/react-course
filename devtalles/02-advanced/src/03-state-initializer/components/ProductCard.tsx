import styles from "../styles/styles.module.css";
import { useProduct } from "../hooks/useProduct";
import { CSSProperties, ReactElement } from "react";
import { Product } from "../data/products";
import { onChangeArgs } from "../hooks/useShoppingCart";
import { Provider } from "../context/ProductContext";
import { ProductCardHandlers } from "../interfaces/interfaces";

// componente ProductCard
export interface InitialValues {
	count: number;
	maxCount: number;
}

export interface ProductCardProps {
	product: Product;
	children: (args: ProductCardHandlers) => ReactElement | ReactElement[];
	className?: string;
	style?: CSSProperties;
	onChange?: (args: onChangeArgs) => void;
	value?: number;
	initialValues?: InitialValues;
}

export const ProductCard = ({
	product,
	children,
	className,
	style,
	onChange,
	value,
	initialValues,
}: ProductCardProps) => {
	const { increaseBy, counter, maxCount, isMaxCountReached, reset } =
		useProduct({
			onChange,
			product,
			value,
			initialValues,
		});
	return (
		<Provider value={{ counter, increaseBy, product, maxCount }}>
			<div className={`${styles.productCard} ${className}`} style={style}>
				{children({
					product,
					maxCount,
					count: counter,
					isMaxCountReached,
					reset,
					increaseBy,
				})}
			</div>
		</Provider>
	);
};
