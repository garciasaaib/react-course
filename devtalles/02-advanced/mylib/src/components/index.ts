import { ProductCard as ProductCardComponent } from "./ProductCard";

import { ProductImage } from "./ProductImage";
import { ProductTitle } from "./ProductTitle";
import { ProductButtons } from "./ProductButtons";
import { ProductCardHOCProps } from "../interfaces/interfaces";

export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";
export { ProductButtons } from "./ProductButtons";

export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardComponent, {
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons,
})

export default ProductCard;