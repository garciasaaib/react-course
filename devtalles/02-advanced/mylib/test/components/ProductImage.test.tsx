import * as React from 'react';
import renderer from 'react-test-renderer';
import { ProductImage } from '../../src/components/ProductImage';
import ProductCard from '../../src/components/index';
import { product1 } from '../../src/data/products';

describe('<ProductTitle />', () => {
  it('renders without crashing', () => {
    const wrapper = renderer.create(
    <ProductImage img='https://image.com' />
  );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should show the component with the product name', () => {
    const wrapper = renderer.create(
    <ProductCard product={product1}>
      {() => (
        <ProductImage />)
      }

    </ProductCard>
  );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
