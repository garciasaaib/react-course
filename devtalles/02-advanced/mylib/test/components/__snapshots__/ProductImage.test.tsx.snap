// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductTitle /> renders without crashing 1`] = `
<img
  alt="Product image"
  className="productImg "
  src="https://image.com"
/>
`;

exports[`<ProductTitle /> should show the component with the product name 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product image"
    className="productImg "
    src="./coffee-mug.png"
  />
</div>
`;
